### ISSUE TITLE
Undo stack implementation is private

### ISSUE THREAD

**Issue by Selmar** [OurMachinery/themachinery-public#815]
I'm trying to use the graph API at runtime. The graph API requires an undo stack implementation, which is implemented in the machinery core and not available for use outside of the editor (the runner does not have an undo stack, of course).

I think it would be nice if the undo stack was part of the public API so it could be used by a plugin, at run-time, without having to copy-paste the code. That being said, I also understand if you believe this falls outside of the scope of what you aim to provide with the public API. I expect there will be more related issues down the road (I haven't taken this very far yet! 😄), so it would be nice to know this for future reference.

**Comment by niklas-ourmachinery** [comment]
Will you make use of the undo functionality in this case? If not, you could just supply a "do nothing" undo stack implementation that doesn't do anything.

We could add something like that to the header for convenience, it would just be a few lines of code.

**Comment by Selmar** [comment]
Yes, we want to provide an editor-like experience to the user. In particular, currently, we intend to expose the graph API with custom nodes, so that the user can use the graphs to express logic for parts of the game. Undo/redo is very welcome there. :)

**Comment by niklas-ourmachinery** [comment]
Sounds reasonable... I think we can provide a default implementation that uses a single "global" undo stack.

(The Machinery uses a per-document undo stack, to have separate undo stacks in separate tabs, but I think that kind of "advanced" usage is more application specific and doesn't belong in the default implementation.)

**Comment by Selmar** [comment]
The simplest would be an undo stack I can instantiate myself, as then I can decide what I pass it to (and re-implement any document-based undo stack myself if needed, though it seems unlikely indeed).

**Comment by niklas-ourmachinery** [comment]
:+1:
